cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/AuthorizationModule.java

```diff
@@ -0,0 +1,172 @@
+    expose(DatasetContext.class);
+
+    bind(Admin.class).toProvider(AdminProvider.class);
+    expose(Admin.class);
+
+    bind(TransactionalFunctionExecutor.class).toProvider(TransactionalFunctionExecutorProvider.class);
+    expose(TransactionalFunctionExecutor.class);
+
+    bind(AuthorizerInstantiatorService.class).in(Scopes.SINGLETON);
+    expose(AuthorizerInstantiatorService.class);
+  }
+
+  private static final class DynamicDatasetCacheProvider implements Provider<DynamicDatasetCache> {
+
+    private final DatasetFramework dsFramework;
+    private final TransactionSystemClient txClient;
+    private final MetricsCollectionService metricsCollectionService;
+
+    @Inject
+    private DynamicDatasetCacheProvider(DatasetFramework dsFramework, TransactionSystemClient txClient,
+                                        MetricsCollectionService metricsCollectionService) {
+      this.dsFramework = dsFramework;
+      this.txClient = txClient;
+      this.metricsCollectionService = metricsCollectionService;
+    }
+
+    @Override
+    public DynamicDatasetCache get() {
+      SystemDatasetInstantiator dsInstantiator = new SystemDatasetInstantiator(dsFramework, null, null);
+      return new MultiThreadDatasetCache(
+        dsInstantiator, txClient, NamespaceId.SYSTEM, ImmutableMap.<String, String>of(),
+        metricsCollectionService.getContext(ImmutableMap.<String, String>of()),
+        ImmutableMap.<String, Map<String, String>>of()
+      );
+    }
+  }
+
+  private static final class DatasetContextProvider implements Provider<DatasetContext> {
+
+    private final DynamicDatasetCache dsCache;
+
+    @Inject
+    private DatasetContextProvider(DynamicDatasetCache dsCache) {
+      this.dsCache = dsCache;
+    }
+
+    @Override
+    public DatasetContext get() {
+      return dsCache;
+    }
+  }
+
+  private static final class AdminProvider implements Provider<Admin> {
+    private final DatasetFramework dsFramework;
+
+    @Inject
+    private AdminProvider(DatasetFramework dsFramework) {
+      this.dsFramework = dsFramework;
+    }
+
+    @Override
+    public Admin get() {
+      return new DefaultAdmin(dsFramework, NamespaceId.SYSTEM);
+    }
+  }
+
+  private static final class TransactionalFunctionExecutorProvider implements Provider<TransactionalFunctionExecutor> {
+
+    private final Logger log = LoggerFactory.getLogger(TransactionalFunctionExecutorProvider.class);
+
+    private final DynamicDatasetCache datasetCache;
+
+    @Inject
+    private TransactionalFunctionExecutorProvider(DynamicDatasetCache datasetCache) {
+      this.datasetCache = datasetCache;
+    }
+
+    @Override
+    public TransactionalFunctionExecutor get() {
+      return new TransactionalFunctionExecutor() {
+        @Override
+        public void execute(TxRunnable runnable) throws TransactionFailureException {
+          TransactionContext transactionContext = datasetCache.get();
```

Comment:
This is duplicated with the other execute method. A execution of `Runnable` can be replaced with a `Callable` (and vice-versa) easily.